### ISSUE TITLE
Detect discrete distributions 

### ISSUE THREAD

**Issue by mjskay** [mjskay/ggdist#19]
And do something sensible with their slab functions (histogram ish things?). May wait on https://github.com/mitchelloharawild/distributional/issues/8

TODOs:
- [x] do this for stat_dist_
  - [x] tests
- ~~[ ] do this for stat_sample_~~ not needed atm, there's always histinterval
- [ ] add example to vignette

**Comment by isabellaghement** [comment]
Hi Matthew, 

Here is the promised R code for displaying discrete conditional Poisson distributions for various values of an X variable.  In the graph, I relabeled the X values as 1 through 5, although they were originally set to something different.  The graph looks like this:
![image](https://user-images.githubusercontent.com/57458528/115442812-50238100-a1c7-11eb-8444-550887999228.png)

The R code is as follows: 

`
library(ggplot2)
theme_set(theme_bw())

beta0 <- 3
beta1 <- -0.5

x <- 1:5 

mu <- exp(beta0 + beta1*x)

mu 

mu <- round(mu)

mu

set.seed(235)
dpois_mu_x_1 <- dpois(0:20, lambda = mu[1])
set.seed(568)
dpois_mu_x_2 <- dpois(0:20, lambda = mu[2])
set.seed(397)
dpois_mu_x_3 <- dpois(0:20, lambda = mu[3])
set.seed(397)
dpois_mu_x_4 <- dpois(0:20, lambda = mu[4])
set.seed(690)
dpois_mu_x_5 <- dpois(0:20, lambda = mu[5])



M <- max(c(dpois_mu_x_1, dpois_mu_x_2, dpois_mu_x_3, 
           dpois_mu_x_4, dpois_mu_x_5))

xM <- c(1, 1 + M, 1 + 2*M, 1 + 3*M, 1 + 4*M)

ggplot(data = NULL, aes(x = xM, y = 1:20)) + 
  geom_segment(aes(x = xM[1], 
                   y = 0:20, 
                   xend = xM[1] + dpois_mu_x_1, 
                   yend = 0:20), size = 1, colour="grey0") + 
  geom_vline(xintercept = xM[1], linetype=2, colour="grey0") + 
  geom_segment(aes(x = xM[2], 
                   y = 0:20, 
                   xend = xM[2] + dpois_mu_x_2, 
                   yend = 0:20), size = 1, colour="grey10") +
  geom_vline(xintercept = xM[2], linetype=2, colour="grey10") + 
  geom_segment(aes(x = xM[3], 
                   y = 0:20, 
                   xend = xM[3] + dpois_mu_x_3, 
                   yend = 0:20), size = 1, colour="grey10") + 
  geom_vline(xintercept = xM[3], linetype=2, colour="grey10") + 
  geom_segment(aes(x = xM[4], 
                   y = 0:20, 
                   xend = xM[4] + dpois_mu_x_4, 
                   yend = 0:20), size = 1, colour="grey10") + 
  geom_vline(xintercept = xM[4], linetype=2, colour="grey10") + 
  geom_segment(aes(x = xM[5], 
                   y = 0:20, 
                   xend = xM[5] + dpois_mu_x_5, 
                   yend = 0:20), size = 1, colour="grey10") +
  geom_vline(xintercept = xM[5], linetype=2, colour="grey10") + 
  coord_cartesian(ylim = c(0,20), xlim = c(xM[1]-0.1, xM[5]+0.3)) + 
  geom_line(aes(x = xM, y = mu), size = 1, colour="darkorange2") + 
  geom_point(aes(x = xM[1], y = mu[1]), size=3.5, colour="darkorange2") + 
  geom_point(aes(x = xM[2], y = mu[2]), size=3.5, colour="darkorange2") + 
  geom_point(aes(x = xM[3], y = mu[3]), size=3.5, colour="darkorange2") +
  geom_point(aes(x = xM[4], y = mu[4]), size=3.5, colour="darkorange2") +
  geom_point(aes(x = xM[5], y = mu[5]), size=3.5, colour="darkorange2") + 
  xlab("X (Predictor Variable)") + 
  ylab("Y (Count Variable)") + 
  scale_x_continuous(breaks = xM, labels = 1:5)
`

Please tag me on Twitter at @IsabellaGhement if you have any questions or comments about the code.

**Comment by mjskay** [comment]
Thanks for this! I'll have to think about what this could look like for the slabinterval family. For those geoms a solution to this would probably come out looking more like a stat_histinterval but with bin sizes automatically set to 1 and centered on each integer. The segment-based approach would probably have to be a new geom altogether.

As for auto-detecting discrete distributions, the only thing I can come up with at the moment is to check if a random draw from the distribution is an integer (I had hoped checking a quantile would work, but all the quantile functions in base seem to return floating point values even if they are from discrete distributions).

**Comment by isabellaghement** [comment]
You're welcome!   If the segments (or bar charts) are used to plot density values associated with integers - as is the case in my plot - I think there might be a "hacky" way to detect that by paying attention to the warnings thrown by R.  

Compare this:

`dpois(c(0,1,2,3), lambda = 2)`

> dpois(c(0,1,2,3), lambda = 2)
[1] 0.1353353 0.2706706 0.2706706 0.1804470

with this:

`dpois(c(0,1.2,2.3,3.4), lambda = 2)`

> dpois(c(0,1.2,2.3,3.4), lambda = 2)
[1] 0.1353353 0.0000000 0.0000000 0.0000000
Warning messages:
1: In dpois(c(0, 1.2, 2.3, 3.4), lambda = 2) : non-integer x = 1.200000
2: In dpois(c(0, 1.2, 2.3, 3.4), lambda = 2) : non-integer x = 2.300000
3: In dpois(c(0, 1.2, 2.3, 3.4), lambda = 2) : non-integer x = 3.400000

So, behind the scenes, you could feed a non-integer value x to the dpois() (or some other discrete density) function as part of the test to see if it's a discrete density, collect the warning produced by R and then search in that (vector of) warning(s) for the phrase _non-integer x_.  If you find that phrase (presuming there is a warning to begin with), you can declare the density as being discrete.  

To collect warnings after calling dpois(), you can try something like this (inspired by the thread https://stackoverflow.com/questions/3903157/how-can-i-check-whether-a-function-call-results-in-a-warning):
 
`
library(purrr)

# create a version of dpois() which will return not just results, but also warnings
qdpois <- quietly(dpois) 

# apply the version of dpois which will return the results and warnings
dens <- qdpois(c(0,1.2,2.3,3.4), lambda = 2)

# access the results
dens$result

# access the warnings 
dens$warnings
`

What comes out from the above code will be something like this: 

> dens
$result
[1] 0.1353353 0.0000000 0.0000000 0.0000000

$output
[1] ""

$warnings
[1] "non-integer x = 1.200000" "non-integer x = 2.300000" "non-integer x = 3.400000"

$messages
character(0)

Then you can pull dens$warnings and check if it has non-zero length first.  If it does have non-zero length, check if any of the warnings include the phrase  "non-integer x" (or even just "non-integer").  If yes, declare dpois to be a discrete distribution. 

This presumes that R throws such a warning any time you try to compute a discrete density for a non-integer value (not sure that is the case - you would need to check this).
  
Another way to verify you are dealing with a discrete distribution would be to check whether the density distribution name (e.g., dpois) falls among a comprehensive list of all discrete density distributions.  

The third way to verify this is to place the onus on the user to declare the distribution type.  For instance, add a parameter like "type" or "distribution_type" to your function and ask the user to set it to "discrete" or "continuous".  Or add a parameter called "discrete" and ask the user to set it to TRUE (for discrete densities) or FALSE (for continuous densities), etc.

Fun programming problem! 😊

**Comment by mjskay** [comment]
I think of the various options, using `is.integer(rdist(1, ...))` for whatever `dist` seems the most reliable to me, as it would be less brittle than parsing error messages. 

Maintaining a list of distributions is an interesting suggestion; if/when {distributional} gets support for a function that returns whether its distributions are discrete that will be an essentially similar solution. Probably the best path is a PR to {distributional} that does that with a fallback to `is.integer(rdist(1, ...))` for unknown distributions.

**Comment by mjskay** [comment]
I've got a first attempt at this working on the `dev` branch currently. To fit in with the existing slab geometry I went with a fill histogram style display rather than a spike histogram. Here's an example similar to yours above:

``` r
library(tidyverse)
library(ggdist)
library(distributional)

tibble(lambda = c(13,7,4,3,2)) %>% 
  ggplot(aes(x = fct_rev(factor(lambda)), dist = dist_poisson(lambda))) + 
  stat_dist_halfeye(interval_color = NA, slab_color = "gray50") + 
  # re-using pointinterval here to get positions for the lines
  stat_dist_pointinterval(geom = "line", .width = .5, size = 1, group = 1)
```

![](https://i.imgur.com/K8DlWBQ.png)

<sup>Created on 2021-07-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by isabellaghement** [comment]
This looks nice but the fill histogram style implies (to me at least) that the Poisson distribution is treated as if it is a _continuous_ distribution rather than a _discrete_ one.  For a discrete distribution, the spike histogram would be more appropriate.    

Just today I came upon a similar figure in Sara Stoudt's slides on GAM models (see figure on the right, which is what we should be aiming for): 

![image](https://user-images.githubusercontent.com/57458528/124330411-377a0480-db42-11eb-8a0c-d05e66f8fc41.png)

The full slides can be accessed here:  https://github.com/SocialScienceDataLab/MZES_GAMs.

**Comment by mjskay** [comment]
I had considered whether or not the histoline version implies the distribution is continuous, and in context I am not convinced it does: as a user you are asking for a density and then getting back a representation that is binned at an integer resolution and is clearly not smooth. That said, I could be wrong :)

I do like the more typical histogram look you suggest though! I should have mentioned that you can use `outline_bars = TRUE` to get that, as with the existing histinterval stat:

```r
library(tidyverse)
library(ggdist)
library(distributional)

tibble(lambda = c(13,7,4,3,2)) %>% 
  ggplot(aes(x = fct_rev(factor(lambda)), dist = dist_poisson(lambda))) + 
  stat_dist_halfeye(interval_color = NA, slab_color = "gray85", outline_bars = TRUE) + 
  # re-using pointinterval here to get positions for the lines
  stat_dist_pointinterval(geom = "line", .width = .5, size = 1, group = 1)
```
![image](https://user-images.githubusercontent.com/6345019/124331501-efb0a880-db54-11eb-8ad0-b1d0a10ba2ce.png)

**Comment by isabellaghement** [comment]
This last plot looks much better to me.  I essentially look at the **footprint** of the histogram - does it imply a continuous range of data values represented by that distribution? 

This one does: 

![image](https://user-images.githubusercontent.com/57458528/124336378-afe8c180-db52-11eb-9624-53d3e3cf637d.png)

This one at least implies there could be a separation between values: 

![image](https://user-images.githubusercontent.com/57458528/124336420-d9095200-db52-11eb-90ea-29d84d490ac5.png)

You seem to look at the **contour** of the histogram, which is a different animal altogether.  To me, the **footprint** conveys information about where the data values actually live (e.g., a continuous range of values; a discrete set of values)  The **contour** conveys information about how many data values actually live in specific portions of that range.

**Comment by mjskay** [comment]
I see your point, but I find the outlines visually noisy without adding much useful information, and I'd rather not have the defaults of this function be different from histinterval anyway (which also defaults to `outline_bars = FALSE`).

The example I included in the vignette inspired by this takes discussion (thanks!) a slightly different approach to emphasizing the discrete nature of the distribution while also showing the use of multiple redundant encodings:

```r
library(tidyverse)
library(ggdist)
library(distributional)

tibble(
  group = c("a","b","c","d","e"),
  lambda = c(13,7,4,3,2)
) %>% 
  ggplot(aes(x = group)) + 
  stat_dist_slab(aes(dist = dist_poisson(lambda), fill = stat(pdf))) + 
  geom_line(aes(y = lambda, group = NA), size = 1) +
  geom_point(aes(y = lambda), size = 2.5) +
  theme_ggdist()
```
![image](https://user-images.githubusercontent.com/6345019/124791001-49c4bb80-df11-11eb-8c62-ed1898f70031.png)

(Yes, the computed variable is called `pdf` not `pmf` since the name `pdf` is already used in stat_dist_slabinterval and changing it for different distribution types would break the generality of the code; I consider this no more egregious than the fact that the pmf of a Poisson in R is called `dpois`).

Thanks for the feedback! Closing this now.
